internal/charmstore/store.go

```diff
@@ -325,29 +344,92 @@ func (s *Store) FindEntity(url *charm.Reference, fields ...string) (*mongodoc.En
 // If any fields are specified, only those fields will be
 // populated in the returned entities.
 func (s *Store) FindEntities(url *charm.Reference, fields ...string) ([]*mongodoc.Entity, error) {
-	var q bson.D
-	if url.Series == "" || url.Revision == -1 {
-		// The url can match several entities - select
-		// based on the base URL and filter afterwards.
-		q = bson.D{{"baseurl", baseURL(url)}}
-	} else {
-		q = bson.D{{"_id", url}}
-	}
-
-	query := selectFields(s.DB.Entities().Find(q), fields)
+	query := selectFields(s.QueryEntity(url), fields)
 	var docs []*mongodoc.Entity
 	err := query.All(&docs)
 	if err != nil {
 		return nil, errgo.Mask(err)
 	}
-	last := 0
-	for _, doc := range docs {
-		if matchURL(doc.URL, url) {
-			docs[last] = doc
-			last++
+	return docs, nil
+}
+
+//FindBestEntity finds the entity that provides the preferred match to
+//the given URL. If any fields are specified, only those fields will be
+//populated in the returned entities.
+func (s *Store) FindBestEntity(url *charm.Reference, fields ...string) (*mongodoc.Entity, error) {
+	if len(fields) > 0 {
+		// make sure we have all the fields we need to make a decision
+		fields = append(fields, "_id", "promulgated-url", "promulgated-revision", "series", "revision")
+	}
+	entities, err := s.FindEntities(url, fields...)
+	if err != nil {
+		return nil, errgo.Mask(err)
+	}
+	if len(entities) == 0 {
+		return nil, errgo.WithCausef(nil, params.ErrNotFound, "entity not found")
+	}
+	best := entities[0]
+	for _, e := range entities {
+		if seriesScore[e.Series] > seriesScore[best.Series] {
+			best = e
+			continue
+		}
+		if seriesScore[e.Series] < seriesScore[best.Series] {
+			continue
+		}
+		if url.User == "" {
+			if e.PromulgatedRevision > best.PromulgatedRevision {
+				best = e
+				continue
+			}
+		} else {
+			if e.Revision > best.Revision {
+				best = e
+				continue
+			}
 		}
 	}
-	return docs[0:last], nil
+	return best, nil
+}
+
+var seriesScore = map[string]int{
+	"bundle":  -1,
+	"lucid":   1000,
+	"precise": 1001,
+	"trusty":  1002,
+	"quantal": 1,
+	"raring":  2,
+	"saucy":   3,
+	"utopic":  4,
+}
+
+//QueryEntity creates a mgo.Query object that can be used to get
+//entities matching the given URL.
+func (s *Store) QueryEntity(url *charm.Reference) *mgo.Query {
```

Comment:
I wonder whether EntityQuery might be better here as it's returning a query, not querying an entity as such. Or QueryEntities perhaps.